Rewrite VecGeom converter to fix mapping issues with reflecting volumes

sethrj opened the pull request:
This is a complete rewrite of the geant4→vecgeom converter for clarity, performance, and accuracy. A few key changes are:
- To more closely reproduce the volume ID layouts for the direct GDML construction, the LogicalVolume mothers are constructed in order based on what's available from Geant4.
- Underlying reflecting volumes are used rather than reflected volumes: I think this is why the old implementation didn't work
- Type IDs are used to dispatch the solid conversion, which improves the performance from linear in number of solids to `O(1)`, *and* catches the case where a user might have derived a shape (unsupported!) from a supported shape
- Instead of replacing an unknown solid with a "wrapper" (which didn't properly scale the underlying volume) we warn and replace with a sphere that has the same radius. This also gives us compatibility with the current VecGeom "master" since v1.2 requires us to define the now-removed vector interfaces.

sethrj committed:
REVERTME: disambiguate _refl

sethrj committed:
Use geo interface to avoid broken include paths

sethrj committed:
Rewrite geant4 geometry converter to fix CMS run3 tracking

sethrj committed:
Revert "REVERTME: disambiguate _refl"

This reverts commit fa107cc729e3b3b5d0a9a7ccd2558668f4c92981.

sethrj committed:
Remove debug message about reflecting volumes, no longer needed

sethrj committed:
Restore volume comparison

whokion commented:
@sethrj Need to add \<memory\> header in celeritas/ext/detail/Converter.hh

sethrj committed:
Fix missing include

sethrj committed:
Change env variable name

sethrj committed:
Print device time log to the same stream as  device status initialization

sethrj committed:
Reset old geometry data before reloading

sethrj committed:
Add test for arbitrary file loading through VGDML

sethrj committed:
Improve time logging

sethrj committed:
Add environment variable for setting vecgeom verbosity

amandalund commented on src/celeritas/ext/VecgeomParams.cc:
```diff
@@ -75,19 +75,25 @@ VecgeomParams::VecgeomParams(G4VPhysicalVolume const* world)
+        g4vg::Converter::Options opts;
+        opts.compare_volumes = [] {
+            std::string var = celeritas::getenv("G4VG_COMPARE_VOLUMES");
+            return !var.empty() || var == "0";
```
Should this be
```suggestion
            return !(var.empty() || var == "0");
```
?

sethrj replied:
Quite so. But since none of the other environment variable conditions (e.g. `CELER_DISABLE_DEVICE`) check for the `== "0"` case I think I'll just kill it entirely.

amandalund commented on src/celeritas/ext/g4vg/LogicalVolumeConverter.hh:
```diff
@@ -0,0 +1,75 @@
+/*!
+ * Convert a Geant4 base LV to a VecGeom LV.
+ *
+ * This does not convert or add any of of the daughters.
```
```suggestion
 * This does not convert or add any of the daughters.
```

amandalund commented on src/celeritas/ext/g4vg/Converter.hh:
```diff
@@ -0,0 +1,111 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/detail/Converter.hh
```
```suggestion
//! \file celeritas/ext/g4vg/Converter.hh
```

amandalund commented on src/celeritas/ext/g4vg/Converter.cc:
```diff
@@ -0,0 +1,207 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/detail/Converter.cc
```
```suggestion
//! \file celeritas/ext/g4vg/Converter.cc
```

amandalund commented on src/celeritas/ext/g4vg/LogicalVolumeConverter.cc:
```diff
@@ -0,0 +1,99 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/detail/LogicalVolumeConverter.cc
```
```suggestion
//! \file celeritas/ext/g4vg/LogicalVolumeConverter.cc
```

amandalund commented on src/celeritas/ext/g4vg/Scaler.hh:
```diff
@@ -0,0 +1,50 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/detail/Scaler.hh
```
```suggestion
//! \file celeritas/ext/g4vg/Scaler.hh
```

amandalund commented on src/celeritas/ext/g4vg/SolidConverter.cc:
```diff
@@ -0,0 +1,648 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/detail/SolidConverter.cc
```
```suggestion
//! \file celeritas/ext/g4vg/SolidConverter.cc
```

amandalund commented on src/celeritas/ext/g4vg/SolidConverter.hh:
```diff
@@ -0,0 +1,120 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/detail/SolidConverter.hh
```
```suggestion
//! \file celeritas/ext/g4vg/SolidConverter.hh
```

amandalund commented on src/celeritas/ext/g4vg/Transformer.hh:
```diff
@@ -0,0 +1,129 @@
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/ext/detail/Transformer.hh
```
```suggestion
//! \file celeritas/ext/g4vg/Transformer.hh
```

amandalund commented on src/celeritas/ext/g4vg/Converter.hh:
```diff
@@ -0,0 +1,111 @@
+
+                                 inline auto Converter::operator()(arg_type)
+                                     -> result_type
+{
```
Missing semicolon -- but there are some addiitional build errors with Geant4 disabled.

amandalund commented on src/celeritas/ext/g4vg/SolidConverter.hh:
```diff
@@ -0,0 +1,120 @@
+    //!@}
+
+  public:
+    inline SolidConverter(Scaler const& convert_scale_,
```
```suggestion
    inline SolidConverter(Scaler const& convert_scale,
```

amandalund commented on src/celeritas/ext/g4vg/SolidConverter.hh:
```diff
@@ -0,0 +1,120 @@
+SolidConverter::SolidConverter(Scaler const& convert_scale_,
+                               Transformer const& convert_transform,
+                               bool compare_volumes)
+    : convert_scale_(convert_scale_)
```
```suggestion
SolidConverter::SolidConverter(Scaler const& convert_scale,
                               Transformer const& convert_transform,
                               bool compare_volumes)
    : convert_scale_(convert_scale)
```

amandalund commented on src/celeritas/ext/g4vg/SolidConverter.cc:
```diff
@@ -0,0 +1,648 @@
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Convert a geant4 solid to a VecGeom solid_base.
```
```suggestion
 * Convert a geant4 solid to a VecGeom "unplaced volume".
```

amandalund reviewed (commented):
Awesome! This looks great @sethrj, thanks for getting this working!

sethrj committed:
Fix documentation and +vecgeom~geant4 build

sethrj committed:
Fix file location

sethrj commented:
Nice catches, thanks @amandalund !

sethrj committed:
Add an option to redirect scoped stream output

sethrj committed:
Replace missing volume with placeholder rather than garbage

This decouples us from the low-level VecGeom implementation, so we can
build with 'master' (vector interfaces removed) as well as with v1.2.2.

sethrj committed:
fixup! Fix file location

amandalund commented on src/celeritas/ext/g4vg/LogicalVolumeConverter.cc:
```diff
@@ -75,11 +76,21 @@ auto LogicalVolumeConverter::make_volume_map() const -> MapLvVolId
+        shape = this->convert_solid_.to_sphere(*g4lv.GetSolid());
+        CELER_LOG(warning)
+            << "Replaced unknown solid with sphere with capacity "
+            << shape->Capacity() << " [cm^3]";
```
Spherical cow?

sethrj replied:
A spherical cow is better than a 10x size cow that crashes when you try to put it on a GPU. 🐮

sethrj committed:
Revert "Add an option to redirect scoped stream output"

This reverts commit 3f4c345bdc2fa0264d8b28a56e835bfc0e8a4b90.

sethrj committed:
Revert "Print device time log to the same stream as  device status initialization"

This reverts commit 1f30ffa6f7624b92aa903bb361dca8a0826256ff.
